hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util;
+
+import java.io.Serializable;
+
+public final class ClassMetadataUtil {
+
+    private ClassMetadataUtil() {
+    }
+
+    public static boolean isClassStaticAndSerializable(Class clazz) {
```

Comment:
You can move these kind of methods in `SerializationUtil`. Specifically, there is already a similar method there - `SerializationUtil#checkSerializable`.

A bigger issue is that this does not cover some serialisation techniques that HZ supports, such as `IdentifiedDataSerializable`, `DataSerializable` and others. For instance (although it's not that common), the user can define a function like this:

```java
public class IDSBiFunction implements BiFunction<String, String, String>, IdentifiedDataSerializable {
    private String returnValue;
    public IDSBiFunction() {
    }

    public IDSBiFunction(String returnValue) {
        this.returnValue = returnValue;
    }

    @Override
    public String apply(String key, String oldValue) {
        return returnValue;
    }

    @Override
    public int getFactoryId() {
        return 1;
    }

    @Override
    public int getClassId() {
        return 0;
    }

    @Override
    public void writeData(ObjectDataOutput out) throws IOException {
        out.writeUTF(returnValue);
    }

    @Override
    public void readData(ObjectDataInput in) throws IOException {
        returnValue = in.readUTF();
    }
}
```

If this class is registered with a factory, such as: `config.getSerializationConfig().addDataSerializableFactory(1, typeId -> new IDSBiFunction());`, it will be possible to serialize and deserialize the class. 

Unfortunately, checking for all possible serialization strategies is probably not going to be easy to do in a utility method. I was initially thinking that we should simply try to serialise the function here but that might incur an unnecessary overhead for most users trying to submit non-serializable functions. So maybe for a best-effort approach we can additionally check if the class inherits `DataSerializable` or `Portable`.  If the class doesn't implement these but is serializable, we can always add more checks in the future.

@cangencer @taburet WDYT?

Reply:
Talking with Marko, seems like the case where the provided instance would implement IDS is something very awkward and usually users would be ok with simple lambdas and `Serializable`. IDS and other serialization techniques would simply execute locally. That means you don't need to change anything. Does that sound ok @cangencer @taburet?

Reply:
It should take BiFunctionEx instead of BiFunction. Then it would be automatically serializable.

Reply:
@cangencer we're talking about overriding `java.util.concurrent.ConcurrentMap#computeIfPresent(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction)`. 

How would it accept `BiFunctionEx`  instead of `BiFunction`? Am I missing something?

Reply:
See the example I posted, it's possible to do it by adding another method and then upcasting. I would also suggest a "checkSerializable" method to fail fast, if something is not serializable: https://github.com/hazelcast/hazelcast-jet/blob/master/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java#L168

Reply:
distributedStream extends Stream -> DistributedStream is analogous to IMap and Stream is analogous to ConcurrentMap

Reply:
Hi @cangencer @mmedenjak ... I pushed a commit to address all feedback received so far. 
I didn't yet add a method to accept a BiFunctionEx. For `computeIfPresent()` implementation, in addition to checking if the BiFunction is Serializable, we also need to check if has a "static" class. So adding the new overloaded method would only have avoided the Serialization check. I thought the performance cost saving would be minimal (I am sure the JVM would also need to do a check or two to figure out which version of the overloaded `computeIfPresent()` to call).
Also, the `SerializationUtil.checkSerializable()` is more robust (in addition to checking Serialization interface, it actually tries to serialize the object, to detect non-serializable fields).
Does this sound okay? Happy to be corrected ...

Reply:
If you have `BiFunctionEx` it should automatically made serializable without having to cast it explicitly. That's the main benefit.

Reply:
I'm fine with the current state, resolving this comment. If @cangencer has other comments, let's open a new thread.